jiant/pytorch_transformers_interface/modules.py

```diff
@@ -0,0 +1,300 @@
+            #  h_lex = self.model.embeddings.dropout(embeddings)
+        else:
+            lex_seq = None  # dummy; should not be accessed.
+
+        if self.embeddings_mode != "only":
+            # encoded_layers is a list of layer activations, each of which is
+            # <float32> [batch_size, seq_len, output_dim]
+            token_types = self.get_seg_ids(ids)
+            output_seq, output_pooled_vec, hidden_states = self.model(
+                ids, token_type_ids=token_types, attention_mask=mask
+            )
+
+        # <float32> [batch_size, var_seq_len, output_dim]
+        return self.prepare_output(output_seq, lex_seq, hidden_states)
+
+
+class XLNetEmbedderModule(PytorchTransformersEmbedderModule):
+    """ Wrapper for XLNet module to fit into jiant APIs. """
+
+    def __init__(self, args):
+
+        super(XLNetEmbedderModule, self).__init__(args)
+
+        self.model = pytorch_transformers.XLNetModel.from_pretrained(
+            args.input_module, cache_dir=self.cache_dir, output_hidden_states=True
+        )
+
+        tokenizer = pytorch_transformers.XLNetTokenizer.from_pretrained(
+            args.input_module, cache_dir=self.cache_dir, do_lower_case="uncased" in args.tokenizer
+        )  # TODO: Speed things up slightly by reusing the previously-loaded tokenizer.
+        self._sep_id = tokenizer.convert_tokens_to_ids("<sep>")
+        self._cls_id = tokenizer.convert_tokens_to_ids("<cls>")
+        self._pad_id = tokenizer.convert_tokens_to_ids("<pad>")
+        self._unk_id = tokenizer.convert_tokens_to_ids("<unk>")
+
+        self.parameter_setup(args)
+
+        self._SEG_ID_CLS = 2
+        self._SEG_ID_SEP = 3
+
+    def apply_boundary_tokens(s1, s2=None):
+        # XLNet-style boundary token marking on string token sequences
+        if s2:
+            return s1 + ["<sep>"] + s2 + ["<sep>", "<cls>"]
+        else:
+            return s1 + ["<sep>", "<cls>"]
+
+    def forward(
+        self, sent: Dict[str, torch.LongTensor], unused_task_name: str = ""
+    ) -> torch.FloatTensor:
+        """ Run XLNet to get hidden states.
+
+        This forward method does preprocessing on the go,
+        changing token IDs from preprocessed word pieces to
+        what AllenNLP indexes.
+
+        Args:
+            sent: batch dictionary
+
+        Returns:
+            h: [batch_size, seq_len, d_emb]
+        """
+        assert "pytorch_transformers_wpm_pretokenized" in sent
+        # <int32> [batch_size, var_seq_len]
+        ids = sent["pytorch_transformers_wpm_pretokenized"]
+        mask = ids != 0
+        # "Correct" ids to account for different indexing between XLNet and
+        # AllenNLP.
+        # The AllenNLP indexer adds a '@@UNKNOWN@@' token to the
+        # beginning of the vocabulary, *and* treats that as index 1 (index 0 is
+        # reserved for padding).
+        ids[ids == 0] = self._pad_id + 2  # Rewrite padding indices.
+        ids[ids == 1] = self._unk_id + 2  # Rewrite UNK indices.
```

Comment:
Do we expect to get AllenNLP `@@UNKNOWN@@` tokens here, instead of OOV words being handled by the sentencepiece model?

For BERT this shouldn't happen and there was an assert:
```
        # Index 1 should never be used since the BERT WPM uses its own
        # unk token, and handles this at the string level before indexing.
        assert (ids > 1).all()
```

Reply:
ids == 1 is the `@@UNKNOWN@@` token, which the model presumably won't know how to handle, so we rewrite it.

For reasons I don't fully understand, we occasionally get `@@UNKNOWN@@` with XLNet but not BERT—presumably this has something to do with the granularity of the smallest word-pieces. I'm erring on the side of leaving the BERT logic as is.

Reply:
Seems worth figuring out if XLNet should have UNKs or not. Every time OpenAI or BERT has UNKs, it's a bug with preprocessing / overlapping experiment directories.

Reply:
I'm not sure I understand why BERT _doesn't_ have UNKs, but it seems pretty clear that XLNet should UNK out rare unicode chars. Here's a mention of that:

https://medium.com/@amanrusia/xlnet-speaks-comparison-to-gpt-2-ea1a4e9ba39e

Reply:
We are getting good (paper-comparable) performance in runs that also produce occasional UNKs for some tasks, which rules out most of the ways I could imagine this going wrong.